### PR TITLE
feat: Add replaceChild and replaceAll methods [WIP]

### PR DESCRIPTION
**Pull request by luanpotter**
# Description

This adds two utility methods, `replaceChild` and `replaceAll`, as per suggested:
image

There are some points to consider about the implementation:

* should we wait for the removal? remove methods are not awaitable currently
* what if we are trying to replace something with itself?
* what if the removal portion fails (component doesn't exist as child, etc)?
* what if the addition portion fails (component already child of someone else, etc)

@erickzanardo what do you think of the questions above, since you had the idea originally?

## Checklist

- [x] I have followed the [Contributor Guide] when preparing my PR.
- [x] I have updated/added tests for ALL new/updated/fixed functionality.
- [x] I have updated/added relevant documentation in `docs` and added dartdoc comments with `///`.
- [x] I have updated/added relevant examples in `examples` or `docs`.

## Breaking Change?

- [ ] Yes, this PR is a breaking change.
- [x] No, this PR is not a breaking change.

[Contributor Guide]: https://github.com/flame-engine/flame/blob/main/CONTRIBUTING.md
[Conventional Commit]: https://conventionalcommits.org
[CHANGELOG]: https://github.com/flame-engine/flame/blob/main/CHANGELOG.md

### CODE REVIEW

File: packages/flame/test/components/component_test.dart [hunk 1]
```diff
@@ -652,6 +652,55 @@ void main() {
       );
     });
 
+    group('Replacing components', () {
```
Comment (luanpotter) [inline packages/flame/test/components/component_test.dart:655]: since we are reusing the remove and add methods, other side-effects and consequences and lifecycle statuses are already tested individually

File: packages/flame/lib/src/components/core/component.dart [hunk 2]
```diff
@@ -619,6 +619,24 @@ class Component {
     parent = newParent;
   }
 
+  /// Replaces a [replaceFrom] child of this component with [replaceWith].
+  /// I.e., removes [replaceFrom], and adds [replaceWith] instead.
+  FutureOr<void> replaceChild(Component replaceFrom, Component replaceWith) {
+    if (replaceFrom == replaceWith) {
+      // nothing to do
+      return Future.value();
+    }
```
Comment (luanpotter) [inline packages/flame/lib/src/components/core/component.dart:628]: could just allow it to trigger any lifecycle hooks
or could throw an error

File: packages/flame/lib/src/components/core/component.dart [hunk 3]
```diff
@@ -619,6 +619,24 @@ class Component {
...
+    remove(replaceFrom);
+    return add(replaceWith);
+  }
+
+  /// Same as [replaceChild], but takes a map of replacements.
+  /// Removes the key and adds the value instead.
+  FutureOr<void> replaceAll(Map<Component, Component> replacements) {
```
Comment (luanpotter) [inline packages/flame/lib/src/components/core/component.dart:635]: the map could contain things like:
* replacements with itself
* replacement cycles (a -> b -> c)
* replacements to the same value
is it worth validating such cases, and if so, what should we do?

File: packages/flame/lib/src/components/core/component.dart (hunk 2 above)
Comment (st-pasha) [reply packages/flame/lib/src/components/core/component.dart:628]: If there is nothing to do, you can simply `return;`

Comment (luanpotter) [reply packages/flame/lib/src/components/core/component.dart:628]: compiler wasn't happy; let me try again

**Comment by st-pasha** [comment]
I believe the `replace()` method is a much needed functionality, I believe I've seen several cases in the past where it might have proved really useful.

However, a simple `..remove(child)..add(newChild)` combo is too trivial. A much more useful functionality would be if the `replace()` method could replace a component without losing its order within the parent's list of children. That is, an actual replacement.

It would be more challenging, of course, since we can't really rely on `priority` field here. But it should be doable, probably needing a new method within the `OrderedSet`.

As for `replaceAll()` method -- its name suggests that it replaces all components within the parent. So why not make it like that? `replaceAll(Iterable<Component> newChildren)` would simply remove all current children and add new ones. This would be simpler to implement than the `replace()`.


> should we wait for the removal? remove methods are not awaitable currently

I'd say the `replace()` method should not be async at all, for several reasons:
- this would create an impression that the returned future resolves once the new component has fully replaced the old one, which isn't true;
- returning such future is possible, it's just `newChild.mounted` -- but in most cases it won't be needed, and in those cases where users might want it, they can just access the `.mounted` future directly.

It is worth pointing out that the future returned from `Component.add()` also suffers from a similar problem: it creates an impression that the future resolves once the component was added, which is not true. Instead, `.add()` returns a future that resolves when the component finishes loading, and even that only sometimes.


> what if we are trying to replace something with itself?

This should probably be a no-op.

> what if the removal portion fails (component doesn't exist as child, etc)?
> what if the addition portion fails (component already child of someone else, etc)

This should produce assertion errors, similarly to how `.remove()` and `.add()` work right now.

**Comment by luanpotter** [comment]
Those are good points @st-pasha, thanks for the thorough review

for the return type: agreed. I will change to void
for the replaceAll: agreed. we can have the Map<to, from> version too and call it something else if @erickzanardo feels passionate about that signature

about the "keeping the order": I agree it would be nice as a user, and I agree it is hard but doable, but I have fears that it might actually cause more confusion than it's worth.
while we could add a "bypass" to ordered_set, it would be kept in a "wrong" state. if anything re-set it, the ordering would suddenly shift.
also it would be weird for unaware users that might for example try to setup a specific priority and it "does't work" because some elements were replaced and had "phantom" priorities.

I think there is a question to be had about getting completely rid of priority, but mixing both priority + a bypass seems more confusing that its worth. wdyt? and also everyone else

**Comment by st-pasha** [comment]
> it would be kept in a "wrong" state. if anything re-set it, the ordering would suddenly shift.

Ah, that's true. We could make a distinction here that when B replaces A and they have the same priority, then B is put exactly where A was. However, if their priorities are different, then we perform the regular `add()` and let the `ComponentSet` deal with the ordering. We could even have a special flag `replace(A, B, {bool copyPriority = false})` which would copy A's priority to B.

There are many cases where the user doesn't explicitly set the priority of components and thus they all remain at 0. At the same time, the relative order of those components still matters, and `replace` should honor it.

> I think there is a question to be had about getting completely rid of priority, but mixing both priority + a bypass seems more confusing that its worth. wdyt?

We had a discussion about this some time ago (#1457), but I don't think we figured out what to do there.

**Comment by erickzanardo** [comment]
> for the replaceAll: agreed. we can have the Map<to, from> version too and call it something else if @erickzanardo feels passionate about that signature

That is a fair point, I agree that `replaceAll` would replace all the current components to the received set.

About having a version that receives a `Map<to, from>`, we could name it `replaceWhere`, wdyt?

**Comment by spydon** [comment]
> About having a version that receives a `Map<to, from>`, we could name it `replaceWhere`, wdyt?

`replaceWhere` sounds more like it should replace if it fulfills some precondition.
Maybe this isn't needed? It would be pretty easy for the user to just iterate over that map and run replace instead?

**Comment by erickzanardo** [comment]
I agree and disagree :P

> `replaceWhere` sounds more like it should replace if it fulfills some precondition. 

Exactly, the map that you pass is they one that sets that condition right, it is a one to one condition clause, but a clause nevertherless?

> Maybe this isn't needed? It would be pretty easy for the user to just iterate over that map and run `replace` instead?

I agree with that hahahha

**Comment by luanpotter** [comment]
Going to close this for now since we never really reached a conclusion and I never followed up. I will open an issue and link this discussion, if anyone is interested to take this on.
